felix/dataplane/linux/ipip_mgr.go

```diff
@@ -86,206 +70,126 @@ type ipipManager struct {
-	case *proto.RouteUpdate:
-		cidr, err := ip.CIDRFromString(msg.Dst)
-		if err != nil {
-			m.logCtx.WithError(err).WithField("msg", msg).
-				Warning("Unable to parse route update destination. Skipping update.")
-			return
-		}
-		if m.ipVersion != cidr.Version() {
-			// Skip since the update is for a mismatched IP version
-			return
-		}
-
-		// In case the route changes type to one we no longer care about...
-		m.deleteRoute(msg.Dst)
-
-		// Process remote IPAM blocks.
-		if isType(msg, proto.RouteType_REMOTE_WORKLOAD) && msg.IpPoolType == proto.IPPoolType_IPIP {
-			m.logCtx.WithField("msg", msg).Debug("IPIP data plane received route update")
-			m.routesByDest[msg.Dst] = msg
-			m.routesDirty = true
-		}
-
-		// Process IPAM blocks that aren't associated to a single or /32 local workload
-		if routeIsLocalBlock(msg, proto.IPPoolType_IPIP) {
-			m.logCtx.WithField("msg", msg).Debug("IPIP data plane received route update for IPAM block")
-			m.localIPAMBlocks[msg.Dst] = msg
-			m.routesDirty = true
-		} else if _, ok := m.localIPAMBlocks[msg.Dst]; ok {
-			m.logCtx.WithField("msg", msg).Debug("IPIP data plane IPAM block changed to something else")
-			delete(m.localIPAMBlocks, msg.Dst)
-			m.routesDirty = true
-		}
-	case *proto.RouteRemove:
-		// Check to make sure that we are dealing with messages of the correct IP version.
-		cidr, err := ip.CIDRFromString(msg.Dst)
-		if err != nil {
-			m.logCtx.WithError(err).WithField("msg", msg).
-				Warning("Unable to parse route removal destination. Skipping update.")
-			return
-		}
-		if m.ipVersion != cidr.Version() {
-			// Skip since the update is for a mismatched IP version
-			return
-		}
-		m.deleteRoute(msg.Dst)
+		activeHostnameToIP: map[string]string{},
+		hostname:           dpConfig.Hostname,
+		tunnelDevice:       tunnelDevice,
+		tunnelDeviceMTU:    mtu,
+		ipVersion:          ipVersion,
+		externalNodeCIDRs:  dpConfig.ExternalNodesCidrs,
+		ipSetDirty:         true,
+		dpConfig:           dpConfig,
+		routeProtocol:      calculateRouteProtocol(dpConfig),
+		logCtx: logrus.WithFields(logrus.Fields{
+			"ipVersion":     ipVersion,
+			"tunnel device": tunnelDevice,
+		}),
+		opRecorder: opRecorder,
+		routeMgr: newRouteManager(
+			mainRouteTable,
+			proto.IPPoolType_IPIP,
+			tunnelDevice,
+			ipVersion,
+			mtu,
+			dpConfig,
+			opRecorder,
+			nlHandle,
+		),
+	}
+
+	m.updateRouteManager()
+	m.mayUpdateRoutes()
+	return m
+}
+
+func (m *ipipManager) updateRouteManager() {
+	m.routeMgr.routeClassTunnel = routetable.RouteClassIPIPTunnel
+	m.routeMgr.routeClassSameSubnet = routetable.RouteClassIPIPSameSubnet
+	m.routeMgr.setTunnelRouteFunc(m.route)
+}
+
+func (m *ipipManager) OnUpdate(protoBufMsg interface{}) {
+	switch msg := protoBufMsg.(type) {
 	case *proto.HostMetadataUpdate:
 		m.logCtx.WithField("hostname", msg.Hostname).Debug("Host update/create")
 		if msg.Hostname == m.hostname {
-			m.setLocalHostAddr(msg.Ipv4Addr)
+			m.routeMgr.updateDataIfaceAddr(msg.Ipv4Addr)
```

Comment:
`routeMgr` has its own `OnUpdate` method - any reason not to just pass all messages directly down by calling `m.routeMgr.OnUpdate(protoBufMsg)`?

Reply:
We can definitely do that. There are 2 reason I did no go that path:
1 - I hoped to process updates only once.
2 - If we want to send `HostMetadataUpadate` down to route manager to let it find its own data interface, then we should do that same thing for vxlan manager by sending `VTEP` updates as well. Then to not mix up, we need to make sure each update is processed by correct version of route manager by checking ip pool.
This can only be done by adding multiple checks if vxlan or ipip tunnel is enabled, making the code less readable.

Does this make sense?

Reply:
Yeah, that sounds reasonable to me - thanks!